piff/gsobject_model.py

```diff
@@ -0,0 +1,389 @@
+from .util import hsm
+
+
+class GSObjectModel(Model):
+    """ Model that takes a fiducial GalSim.GSObject and dilates, shifts, and shears it to get a
+    good match to stars.
+
+    :param gsobj:    GSObject to use as fiducial profile.
+    :param fastfit:  Use HSM moments for fitting.  Approximate, but fast.  [default: False]
+    :param force_model_center: If True, PSF model centroid is fixed at origin and
+                        PSF fitting will marginalize over stellar position.  If False, stellar
+                        position is fixed at input value and the fitted PSF may be off-center.
+                        [default: True]
+    :param include_pixel:   Include integration over pixel?  [default: True]
+    :param logger:   A logger object for logging debug info. [default: None]
+    """
+    def __init__(self, gsobj, fastfit=False, force_model_center=True, include_pixel=True,
+                 logger=None):
+        import galsim
+        if isinstance(gsobj, basestring):
+            gsobj = eval(gsobj)
+
+        self.kwargs = {'gsobj':repr(gsobj),
+                       'fastfit':fastfit,
+                       'force_model_center':force_model_center,
+                       'include_pixel':include_pixel}
+
+        self.gsobj = gsobj
+        self._fastfit = fastfit
+        self._force_model_center = force_model_center
+        self._method = 'auto' if include_pixel else 'no_pixel'
+        # Params are [u, v], size, g1, g2
+        if self._force_model_center:
+            self._nparams = 3
+            params = [1.0, 0.0, 0.0]
+        else:
+            self._nparams = 5
+            params = [0.0, 0.0, 1.0, 0.0, 0.0]
+
+        # Calibrate gsobj by measuring it with HSM.  This way we can use differences in HSM moments
+        # to get a reasonable starting guess for stars.
+        prof = self.getProfile(params)
+        img = prof.drawImage(method=self._method)
+        sd = StarData(img, img.trueCenter())
+        fiducial_star = Star(sd, None)
+        flux, cenu, cenv, size, g1, g2, flag = hsm(fiducial_star)
+        if flag != 0:
+            raise RuntimeError("Error calculating fiducial moments for this gsobject.")
+        self._fiducial_size = size
+        self._fiducial_shape = galsim.Shear(g1=g1, g2=g2)
+        self._fiducial_flux = flux
+        self._fiducial_centroid = galsim.PositionD(cenu, cenv)
+
+    def moment_fit(self, star, logger=None):
+        """Estimate transformations needed to bring self.gsobj towards given star."""
+        import galsim
+        flux, cenu, cenv, size, g1, g2, flag = hsm(star)
+        centroid = galsim.PositionD(cenu, cenv)
+        shape = galsim.Shear(g1=g1, g2=g2)
+
+        if flag:
+            raise RuntimeError("Failed measuring HSM moments of star.")
+
+        dflux = flux/self._fiducial_flux
+        dcentroid = centroid - self._fiducial_centroid
+        dsize = size/self._fiducial_size
+        dshape = shape - self._fiducial_shape
```

Comment:
I had another idea about this on the drive home.  I think probably what should belong to the star is the hsm measurement of the data.  Then I think this is a reasonable flow for repeated calls to `fit`:
1. First pass checks for the stored hsm measurement, sees that it isn't there, and calculates it.  It compares to the fiducial stuff in the class to derive estimates of size, flux, etc.  i.e. the same thing that is currently being done.
2+. Later passes see that there is already a stored hsm result, which means that it should also have a `fit.params` with the current estimate of the size, shape, etc.  So now, rather than run hsm of the data, it draws the current model using the same image size, scale, etc. and runs hsm on that.  Then it can refine the estimate of the params by comparing the two hsm results.

A couple thoughts about this proposed process.  First I'm not sure where to store the hsm results in the star object.  I guess star.data.properties, but I haven't thought through whether there are any consequences to that.  I think it should be fine though.

Second, I think a few iterations of this may be as accurate as the lmfit stuff.  If so, since we will typically run through a few passes anyway to deal with outliers, it's possible this could just become the one implementation of fit in this class, rather than use the lmfit stuff.  (OTOH, it's also possible that it won't.  Just it's worth checking to see.)